### PR TITLE
AP-306 adjust currencycloud client

### PR DESCRIPTION
**Pull request by oleg-sviridov**
Adjust currencycloud client to be scoped and work with injected http clients and added scoped authorization service with singleton token holder. Token is updated in a synchronized manner using semaphoreSlim instance

### CODE REVIEW

**Comment by vsevolod-mozgovoi-am** [comment]
@oleg-sviridov conflicts

File: Source/CurrencyCloud/Authorization/AuthorizationOptions.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    public class AuthorizationOptions
+    {
+        public AuthorizationOptions(Credentials credentials, int tokenInactivityTimeoutInMinutes)
+        {
+            Credentials = credentials;
+            TokenInactivityTimeoutInMinutes = TimeSpan.FromMinutes(tokenInactivityTimeoutInMinutes);
+        }
+
+        public Credentials Credentials { get; }
+        public TimeSpan TokenInactivityTimeoutInMinutes { get; }
```
Comment (maksim-resnenko) [inline Source/CurrencyCloud/Authorization/AuthorizationOptions.cs:13]: Looks better if you just use `TokenInactivityTimeout`

File: Source/CurrencyCloud/Authorization/AuthorizationOptions.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿namespace CurrencyCloud.Authorization
+{
+
+    public class AuthorizationOptions
+    {
+        public AuthorizationOptions(Credentials credentials, int tokenInactivityTimeoutInMinutes)
```
Comment (maksim-resnenko) [inline Source/CurrencyCloud/Authorization/AuthorizationOptions.cs:6]: TimeSpan TokenInactivityTimeout

File: Source/CurrencyCloud/Client.cs [hunk 3]
```diff
@@ -280,31 +264,22 @@ public async Task OnBehalfOf(string id, Func<Task> function)
...
         /// Gets a value that indicates whether the client is initialized.
         /// </summary>
-        public bool IsInitialized
-        {
-            get
-            {
-                return httpClient != null;
-            }
-        }
+        public bool IsInitialized => httpClient != null && httpClient.DefaultRequestHeaders.Contains("X-Auth-Token");
```
Comment (maksim-resnenko) [inline Source/CurrencyCloud/Client.cs:267]: It seems that `httpClient `will always be non-null, you check in the constructor

File: Source/CurrencyCloud/Authorization/AuthorizationService.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
+﻿using Newtonsoft.Json.Linq;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Net;
+
+namespace CurrencyCloud.Authorization;
+public class AuthorizationService : IAuthorizationService
+
```
Comment (maksim-resnenko) [inline Source/CurrencyCloud/Authorization/AuthorizationService.cs:8]: remove

**Review by maksim-resnenko — APPROVED** [review]

File: Source/CurrencyCloud/Authorization/AuthorizationOptions.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿namespace CurrencyCloud.Authorization
+{
+
```
Comment (vsevolod-mozgovoi-am) [inline Source/CurrencyCloud/Authorization/AuthorizationOptions.cs:3]: redundant

File: Source/CurrencyCloud/Client.cs [hunk 6]
```diff
@@ -214,10 +194,15 @@ await retryPolicy.ExecuteAsync(
...
                 try
                 {
-                    if (attempts > 0)
+                    if (attempts == 1)
                     {
                         await AuthorizeAsync();
                     }
+                    else if (attempts > 1)
+                    {
+                        await AuthorizeAsync(true);
```
Comment (vsevolod-mozgovoi-am) [inline Source/CurrencyCloud/Client.cs:203]: named parameter required

**Review by vsevolod-mozgovoi-am — APPROVED** [review]

File: Source/CurrencyCloud/Client.cs (hunk 6 above)
Comment (oleg-sviridov) [reply Source/CurrencyCloud/Client.cs:203]: there are not format rules for this project, but anyway I thought this rules only applies to cases when passing null?

**Review by baklanser — APPROVED** [review]

**Pull request merged**